lib/windowsexec/syscall_windows.go

```diff
@@ -0,0 +1,150 @@
+	nShow          int
+	hInstApp       windows.Handle
+	lpIDList       uintptr
+	lpClass        uintptr
+	hkeyClass      windows.Handle
+	dwHotKey       uint32
+	hIconOrMonitor windows.Handle
+	hProcess       windows.Handle
+}
+
+// These consts are copied verbatim from
+// https://learn.microsoft.com/en-us/windows/win32/api/shellapi/ns-shellapi-shellexecuteinfow
+const (
+	// SEE_MASK_NOCLOSEPROCESS (0x00000040):
+	// Use to indicate that the hProcess member receives the process handle.
+	// This handle is typically used to allow an application to find out when a
+	// process created with ShellExecuteEx terminates. In some cases, such as
+	// when execution is satisfied through a DDE conversation, no handle will be
+	// returned. The calling application is responsible for closing the handle
+	// when it is no longer needed.
+	SEE_MASK_NOCLOSEPROCESS = 0x40
+)
+
+// RunAsAndWait uses `ShellExecuteExW` to create a new process with elevated
+// privileges on Windows. It waits for the process to exit, or until timeout,
+// is exhausted. It will return an error if the process exits with a non-zero
+// status code.
+func RunAsAndWait(
+	file, directory string, timeout time.Duration, parameters []string,
+) error {
+	// Convert our various string inputs to UTF16Ptrs
+	lpVerb, err := windows.UTF16PtrFromString("runas")
+	if err != nil {
+		return trace.Wrap(err, "converting verb to ptr")
+	}
+	lpFile, err := windows.UTF16PtrFromString(file)
+	if err != nil {
+		return trace.Wrap(err, "converting file to ptr")
+	}
+	lpDirectory, err := windows.UTF16PtrFromString(directory)
+	if err != nil {
+		return trace.Wrap(err, "converting directory to ptr")
+	}
+	lpParameters, err := windows.UTF16PtrFromString(strings.Join(parameters, " "))
+	if err != nil {
+		return trace.Wrap(err, "converting parameters to ptr")
+	}
+
+	// https://learn.microsoft.com/en-us/windows/win32/api/shellapi/nf-shellapi-shellexecuteexw
+	// Results are returned back into info :)
+	info := &shellExecuteInfoW{
+		fMask:        SEE_MASK_NOCLOSEPROCESS,
+		lpVerb:       uintptr(unsafe.Pointer(lpVerb)),
+		lpFile:       uintptr(unsafe.Pointer(lpFile)),
+		lpParameters: uintptr(unsafe.Pointer(lpParameters)),
+		lpDirectory:  uintptr(unsafe.Pointer(lpDirectory)),
+		nShow:        windows.SW_NORMAL,
+	}
+	// Set the size field of info to the size of info.
+	info.cbSize = uint32(unsafe.Sizeof(*info))
+	err = shellExecuteExW(info)
+	if err != nil {
+		// Errors from this can be a little vague, and the contents of hInstApp
+		// can provide additional context. We'll emit a debug log so this is a
+		// little easier to investigate if a user experience issues with this.
+		log.WithFields(log.Fields{
+			"err":      err,
+			"hInstApp": info.hInstApp,
+		}).Debug("Encountered error calling shellExecuteExW")
+		return trace.Wrap(err, "calling shellExecuteExW")
+	}
+	// Since we provided SEE_MASK_NOCLOSEPROCESS, closing info.hProcess is our
+	// responsibility.
+	defer windows.CloseHandle(info.hProcess)
```

Comment:
The docs for `ShellExecuteEx` make it seem like in some edge cases `hProcess` may be NULL even when no error is returned, so I'd sanity check that it isn't `nil` before using it just in case. Windows has so many edge cases...